### PR TITLE
ProductUrlProvider: iterate only over channel locales

### PR DESCRIPTION
**Pull request by SebLours**
With this PR only products available in channel locales are provided.

### CODE REVIEW

File: src/Provider/ProductUrlProvider.php [hunk 1]
```diff
@@ -84,13 +86,22 @@ public function getName(): string
      */
     public function generate(): iterable
     {
+        /** @var ChannelInterface $channel */
+        $channel = $this->channelContext->getChannel();
+
+        Assert::isInstanceOf($channel, ChannelInterface::class);
```
[the remark this answers is not in the document]
Comment (SebLours) [reply src/Provider/ProductUrlProvider.php]: `Sylius\Component\Channel\Context\ChannelContextInterface` returns an instance of `Sylius\Component\Channel\Model\ChannelInterface`.
I just want to be sure that the returned channel is an instance of `Sylius\Component\Core\Model\ChannelInterface` to have the channel locales support.
If you think it isn't necessary i can remove it.

File: src/Provider/ProductUrlProvider.php [hunk 2]
```diff
@@ -84,13 +86,22 @@ public function getName(): string
...
+        $locales = $channel->getLocales();
+
         foreach ($this->getProducts() as $product) {
             $productUrl = $this->sitemapUrlFactory->createNew();
             $productUrl->setChangeFrequency(ChangeFrequency::always());
             $productUrl->setPriority(0.5);
             $productUrl->setLastModification($product->getUpdatedAt());
 
-            foreach ($product->getTranslations() as $translation) {
+            foreach ($locales as $locale) {
```
[the remark this answers is not in the document]
Comment (SebLours) [reply src/Provider/ProductUrlProvider.php]: Ok i'm going to fix that.

[the remark this answers is not in the document]
Comment (stefandoorn) [reply src/Provider/ProductUrlProvider.php]: Great!

File: src/Provider/ProductUrlProvider.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefandoorn) [reply src/Provider/ProductUrlProvider.php]: Yeah.. not sure about it. I think that Sylius cannot work without a channel. Let's try without the assertion first.

[the remark this answers is not in the document]
Comment (SebLours) [reply src/Provider/ProductUrlProvider.php]: Fixed  :sunglasses:

File: src/Provider/ProductUrlProvider.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (SebLours) [reply src/Provider/ProductUrlProvider.php]: Fixed  :sunglasses:

**Comment by stefandoorn** [comment]
Thanks for the updates :) 

In the AbstractTestController now also the FR channel is added, and therefore these tests pass. Would you be able to keep e.g. the old tests intact and add extra tests that cover this specific case? Reason is that in that way we cover more situations:

* The situation where all the locales are present in the channel
* The situation where not all the locales are present in the channel

Sames goes for the specs I think. I think we could extend them in the same manner. It probably feels a bit overdone, but better safe than sorry imo.

**Comment by SebLours** [comment]
I've fixed the tests.
Products have two translations (`en_US`, `nl_NL`), and tests cover this cases:
- channel with only `en_US`
- channel with `en_US`, `nl_NL`

That's the right way for you?

**Comment by stefandoorn** [comment]
Thanks @SebLours! I've rebased and refined in https://github.com/stefandoorn/sitemap-plugin/pull/37. It was quite a work, because I didn't spot soon enough that we added NL as default second locale in the test suite, so the XML output was different.

**Pull request closed without merging**